cardano-testnet/src/Testnet/Components/Query.hs

```diff
@@ -98,25 +106,111 @@ waitForEpochs epochStateView@EpochStateView{nodeConfigPath, socketPath} interval
   currentEpoch <- getCurrentEpochNo epochStateView
   waitUntilEpoch nodeConfigPath socketPath $ addEpochInterval currentEpoch interval
 
+
+-- | Wait for the requested number of blocks
+waitForBlocks
+  :: HasCallStack
+  => MonadIO m
+  => MonadTest m
+  => EpochStateView
+  -> Word64 -- ^ Number of blocks to wait
+  -> m BlockNo -- ^ The block number reached
+waitForBlocks epochStateView numberOfBlocks = withFrozenCallStack $ do
```

Comment:
Couldn't this be implemented more simply in terms of  `watchEpochStateUpdate`? Is there a reason not to?

Reply:
It uses `watchEpochStateUpdate`. Can you explain what do you mean?

Reply:
Ah, sorry, you are right, I must have assumed it wasn't because of the `State` monad.

So why the `State` monad then? Is it for it to be a single call because you don't know the current block?

Still, I think it would be nicer to have it like `waitTillBlock (getBlockNumber + numberOfBlocks)`.

Having state adds complexity, so it makes it more error prone and harder to understand.

Reply:
Also, it would be possible to have `watchEpochStateUpdate` use the latest `block` straight away first (instead of start by waiting for the next `block` after it is called to be produced), that way we wouldn't be using one block just for finding out the block number

Reply:
Fair point. `State` is not really needed here. I'll simplify this function.